### ISSUE TITLE
"ls" takes a very long time when using NFS on JuiceFS

### ISSUE THREAD

**Issue by huntersman** [juicedata/juicefs#2594]
**What happened**:
Test `ls` performace on NFS. It takes about 10 minutes for 20000 files.
**What you expected to happen**:
Less than 1 second for 20000 files.
**How to reproduce it (as minimally and precisely as possible)**:
### JuiceFS and NFS server
```bash
juicefs format \
    --storage s3 \
    --bucket http://serverIp/bucketName \
    --access-key ak \
    --secret-key sk \
    "mysql://user:password@(127.0.0.1:3306)/jfs" \
    myjfs
```
```bash
juicefs mount -d "mysql://user:password@(127.0.0.1:3306)/jfs" /mnt/jfs
```
```bash
yum -y install nfs-utils rpcbind
```
```bash
vi /etc/exports
/mnt/jfs *(rw,insecure,async,no_subtree_check,no_root_squash,fsid=0)

service rpcbind start
service nfs start
```
### NFS client
```bash
mount -t nfs -o nfsvers=3 -o sync serverIp:/mnt/jfs /mnt/nfs
```
Add 20000 4KB files into /mnt/nfs
```bash
seq 20000|xargs -i dd if=/dev/zero of=/mnt/nfs/testfile_{}.cache bs=4k count=1
```
Test `ls` performance and it takes about 10 minutes.
```bash
time ls /mnt/nfs
real 10m36.130s
user 0m0.105s
sys  0m0.378s
```
**Anything else we need to know?**
I use `juicefs stats /mnt/jfs` and notice that cpu usage becomes really high for a long time.

BTW, I test only use JuiceFS or NFS, both performance are good.
```bash
# 20000 4KB files
time ls /mnt/jfs
real 0m0.335s
user 0m0.054s
sys  0m0.031s
```
**Environment**:
- JuiceFS version (use `juicefs --version`) or Hadoop Java SDK version: juicefs version 1.0.0+2022-08-08.cf0c269
- OS (e.g `cat /etc/os-release`): CentOS 7
- Kernel (e.g. `uname -a`): 5.4.210-1.el7.elrepo.x86_64
- Object storage (cloud provider and region, or self maintained): self maintained
- Metadata engine info (version, cloud provider managed or self maintained): self maintained, MySQL 5.7.33

Thanks for your help!

**Comment by SandyXSD** [comment]
When you call `ls` directly on JuiceFS, the entries are retrieved only once. But when you do that on NFS client, things become complicated. The NFS client will issue a lot of `readdir` rpc calls sequentially, each of which has limited buffer for about 100 entries. For every rpc request, JuiceFS (NFSD) runs an `opendir -> readdir -> readir -> ... -> closedir` procedure, retrieving all entries even though it returns only a small subset. The entry cache is dropped after directory closed, so JuiceFS has to run the procedure again for the next request. As a result, JuiceFS is readding all the entries over and over again, consuming a lot of CPU as you observed.
One possible fix is to keep entry cache for a while even after directory is closed, but we won't do that for now as it may cause other issues.

Some access logs are as followings:

```
2022.08.24 11:23:42.201737 [uid:0,gid:0,pid:26290] opendir (2): OK [fh:2242] <0.000003>
2022.08.24 11:23:42.207992 [uid:0,gid:0,pid:0] releasedir (2): OK <0.000001>
2022.08.24 11:23:42.626464 [uid:0,gid:0,pid:26290] readdir (2,4096,612): OK (19390) <0.424694>
2022.08.24 11:23:42.626579 [uid:0,gid:0,pid:26290] readdir (2,4096,635): OK (19367) <0.000002>
2022.08.24 11:23:42.626641 [uid:0,gid:0,pid:26290] readdir (2,4096,658): OK (19344) <0.000001>
2022.08.24 11:23:42.626704 [uid:0,gid:0,pid:26290] readdir (2,4096,681): OK (19321) <0.000000>
2022.08.24 11:23:42.626756 [uid:0,gid:0,pid:26290] readdir (2,4096,704): OK (19298) <0.000000>
2022.08.24 11:23:42.626806 [uid:0,gid:0,pid:26290] readdir (2,4096,727): OK (19275) <0.000001>
2022.08.24 11:23:42.626856 [uid:0,gid:0,pid:26290] readdir (2,4096,750): OK (19252) <0.000000>
2022.08.24 11:23:42.626910 [uid:0,gid:0,pid:26290] readdir (2,4096,773): OK (19229) <0.000000>
2022.08.24 11:23:42.626958 [uid:0,gid:0,pid:26290] readdir (2,4096,796): OK (19206) <0.000000>
2022.08.24 11:23:42.627203 [uid:0,gid:0,pid:26290] getattr (2): OK (2,[drwxr-xr-x:0040755,2,0,0,1661332257,1661333040,1661333040,4096]) <0.0001
2022.08.24 11:23:42.627548 [uid:0,gid:0,pid:26290] opendir (2): OK [fh:2243] <0.000002>
2022.08.24 11:23:42.635990 [uid:0,gid:0,pid:0] releasedir (2): OK <0.000001>
2022.08.24 11:23:43.051167 [uid:0,gid:0,pid:26290] readdir (2,4096,714): OK (19288) <0.423584>
2022.08.24 11:23:43.051295 [uid:0,gid:0,pid:26290] readdir (2,4096,737): OK (19265) <0.000002>
2022.08.24 11:23:43.051372 [uid:0,gid:0,pid:26290] readdir (2,4096,760): OK (19242) <0.000001>
2022.08.24 11:23:43.051427 [uid:0,gid:0,pid:26290] readdir (2,4096,783): OK (19219) <0.000001>
2022.08.24 11:23:43.051482 [uid:0,gid:0,pid:26290] readdir (2,4096,806): OK (19196) <0.000001>
2022.08.24 11:23:43.051533 [uid:0,gid:0,pid:26290] readdir (2,4096,829): OK (19173) <0.000000>
2022.08.24 11:23:43.051582 [uid:0,gid:0,pid:26290] readdir (2,4096,852): OK (19150) <0.000000>
2022.08.24 11:23:43.051634 [uid:0,gid:0,pid:26290] readdir (2,4096,875): OK (19127) <0.000000>
2022.08.24 11:23:43.051684 [uid:0,gid:0,pid:26290] readdir (2,4096,898): OK (19104) <0.000000>
```

**Comment by huntersman** [comment]
> When you call `ls` directly on JuiceFS, the entries are retrieved only once. But when you do that on NFS client, things become complicated. The NFS client will issue a lot of `readdir` rpc calls sequentially, each of which has limited buffer for about 100 entries. For every rpc request, JuiceFS (NFSD) runs an `opendir -> readdir -> readir -> ... -> closedir` procedure, retrieving all entries even though it returns only a small subset. The entry cache is dropped after directory closed, so JuiceFS has to run the procedure again for the next request. As a result, JuiceFS is readding all the entries over and over again, consuming a lot of CPU as you observed. One possible fix is to keep entry cache for a while even after directory is closed, but we won't do that for now as it may cause other issues.
> 
> Some access logs are as followings:
> 
> ```
> 2022.08.24 11:23:42.201737 [uid:0,gid:0,pid:26290] opendir (2): OK [fh:2242] <0.000003>
> 2022.08.24 11:23:42.207992 [uid:0,gid:0,pid:0] releasedir (2): OK <0.000001>
> 2022.08.24 11:23:42.626464 [uid:0,gid:0,pid:26290] readdir (2,4096,612): OK (19390) <0.424694>
> 2022.08.24 11:23:42.626579 [uid:0,gid:0,pid:26290] readdir (2,4096,635): OK (19367) <0.000002>
> 2022.08.24 11:23:42.626641 [uid:0,gid:0,pid:26290] readdir (2,4096,658): OK (19344) <0.000001>
> 2022.08.24 11:23:42.626704 [uid:0,gid:0,pid:26290] readdir (2,4096,681): OK (19321) <0.000000>
> 2022.08.24 11:23:42.626756 [uid:0,gid:0,pid:26290] readdir (2,4096,704): OK (19298) <0.000000>
> 2022.08.24 11:23:42.626806 [uid:0,gid:0,pid:26290] readdir (2,4096,727): OK (19275) <0.000001>
> 2022.08.24 11:23:42.626856 [uid:0,gid:0,pid:26290] readdir (2,4096,750): OK (19252) <0.000000>
> 2022.08.24 11:23:42.626910 [uid:0,gid:0,pid:26290] readdir (2,4096,773): OK (19229) <0.000000>
> 2022.08.24 11:23:42.626958 [uid:0,gid:0,pid:26290] readdir (2,4096,796): OK (19206) <0.000000>
> 2022.08.24 11:23:42.627203 [uid:0,gid:0,pid:26290] getattr (2): OK (2,[drwxr-xr-x:0040755,2,0,0,1661332257,1661333040,1661333040,4096]) <0.0001
> 2022.08.24 11:23:42.627548 [uid:0,gid:0,pid:26290] opendir (2): OK [fh:2243] <0.000002>
> 2022.08.24 11:23:42.635990 [uid:0,gid:0,pid:0] releasedir (2): OK <0.000001>
> 2022.08.24 11:23:43.051167 [uid:0,gid:0,pid:26290] readdir (2,4096,714): OK (19288) <0.423584>
> 2022.08.24 11:23:43.051295 [uid:0,gid:0,pid:26290] readdir (2,4096,737): OK (19265) <0.000002>
> 2022.08.24 11:23:43.051372 [uid:0,gid:0,pid:26290] readdir (2,4096,760): OK (19242) <0.000001>
> 2022.08.24 11:23:43.051427 [uid:0,gid:0,pid:26290] readdir (2,4096,783): OK (19219) <0.000001>
> 2022.08.24 11:23:43.051482 [uid:0,gid:0,pid:26290] readdir (2,4096,806): OK (19196) <0.000001>
> 2022.08.24 11:23:43.051533 [uid:0,gid:0,pid:26290] readdir (2,4096,829): OK (19173) <0.000000>
> 2022.08.24 11:23:43.051582 [uid:0,gid:0,pid:26290] readdir (2,4096,852): OK (19150) <0.000000>
> 2022.08.24 11:23:43.051634 [uid:0,gid:0,pid:26290] readdir (2,4096,875): OK (19127) <0.000000>
> 2022.08.24 11:23:43.051684 [uid:0,gid:0,pid:26290] readdir (2,4096,898): OK (19104) <0.000000>
> ```

Thank you, I understand. It seems too hard to fix it. Maybe the best solution is not using NFS on JuiceFS :joy:
